### ISSUE TITLE
Simple nlu confidence evaluation

### ISSUE THREAD

**Issue by dsjaai** [RasaHQ/rasa#6270]
**Description of Problem**: One goal of a chatbot is to understand the users uttered intention. To be able to consistently identify whether the nlu interpreter understood or evt. misunderstood the uttered intention, the nlu confidence for correct predictions should be high and for incorrect predictions should be low. The current nlu.test does not evaluate the predicted confidences. In a similar fashion this issue was also adressed in:

*   [forum: set-threshold-when-evaluation-a-rasa-nlu](https://forum.rasa.com/t/set-threshold-when-evaluation-a-rasa-nlu/25826)
*   [feature request: 5988](https://github.com/RasaHQ/rasa/issues/5988)

**Overview of the Solution**:

1.  It would be nice to have some visual evaluation of a single model performance and the impact of different nlu thresholds and ambigutiy thresholds (the nlu confidence histogram is not enough).
2.  To be able to measure the performance of a nlu model not only the f1-score should be taken into account but also the separability of wrong predictions and correct predictions using the confidence results. Therefore a metric should be introduced. Such a metric would also help to perform hyperparameter optimization automatically.

**Examples**:

1.  Visualization of the effect of different confidence thresholds: _(To include the ambiguity threshold the nlu.test should output intent_ranking for errors and successes as well)_ For instance, several figures with different y-axis and line graphs:

*   x-axis is the nlu_threshold
*   One curve for each ambiguity_threshold to be tested.
*   y-axis could be: 
      *   version 1: P(fallback triggered=confidence < nlu_threshold OR ambiguous prediction)
    *   version 2: P(correct prediction AND no fallback) + P(wrong prediction AND fallback triggered)

Version 1: example figure [<span class="image-wrap" style="">![](https://user-images.githubusercontent.com/68738846/88400294-6e8b5600-cdc8-11ea-8686-9d150acf8201.png)</span>](https://user-images.githubusercontent.com/68738846/88400294-6e8b5600-cdc8-11ea-8686-9d150acf8201.png)

Version 2: example figure [<span class="image-wrap" style="">![](https://user-images.githubusercontent.com/68738846/88400267-64695780-cdc8-11ea-8b2a-a7f4ddee9fa3.png)</span>](https://user-images.githubusercontent.com/68738846/88400267-64695780-cdc8-11ea-8b2a-a7f4ddee9fa3.png)

1.  Confidence Metric

*   For the [nlu-hyperopt](https://github.com/RasaHQ/nlu-hyperopt) project a loss function or evaluation metric was already introduced. I like the approach, however it requires to include a nlu threshold before training and does not take into account the ambiguity threshold.
*   A metric which is independent from the actual nlu_threshold and ambiguity_threshold would be the following: Its goal is to compute how well the distribution of the confidence for all wrong predictions is different from the distribution of the confidence for all correct predictions while of course confidence for correct predictions should be higher than for wrong predictions: 1) Compute the X%_percentile of the confidence for all wrong predictions and the Z%_percentile for all correct predictions for all X,Z in {5,10,....,90,95} 

    . Then 2) count how many comparisons are true of X%_percentile < Z%_percentile for all X and Z. 3) Divide by all possible comparisons to normalize the result. **Higher values indicate higher separability of wrong and correct predictions based by nlu confidence**

**Blockers** (if relevant):

*   None are known

**Future Ideas** If this feature is appreciated, it maybe should be provided for entities, responses (ResponseSelector) as well?

**Definition of Done**:

*   [ ] Visualization of the effect of different confidence thresholds added
*   [ ] At least one metric for the confidence distribution of correct and wrong nlu predictions are outputted by the nlu.test 
      *   in a file AND in in the log
*   [ ] Tests are added
*   [ ] Feature described the docs
*   [ ] Feature mentioned in the changelog

**Comment by sara-tagger** [comment]
Thanks for submitting this feature request 🚀 @JustinaPetr will get back to you about it soon! ✨

**Comment by Asma-droid** [comment]
hi @dsjaa,

I have appreciated the idea about hyperopt but i dont now how to integrate it in my docker-compose

version: '3.1'
services:
  rasa:
    image: ..........
    container_name: rasa
    networks: ['rasa-network']
    ports:
    - "5005:5005"
    environment:
      rasa_endpoint: "http://localhost:5005"
    volumes:
    - "./models:/app/models"
    - "./config:/app/config"
    command: "run --enable-api --debug --credentials /app/config/credentials.yml --endpoints /app/config/endpoints.yml"

  action_server:
    image: ..........
    networks: ['rasa-network']
    ports:
    - "5055:5055"
    depends_on:
    - mongo
    links:
    - mongo
    volumes:
    - "./actions:/app/actions"


  mongo:
    image: mongo
    restart: always
    container_name: mongo
    networks: ['rasa-network']
    volumes:
      - "mongodbdata:/data/db"
    environment:
      MONGODB_DATABASE: rasaprod
    expose:
     - "27017"
    ports:
      - 27017:27017


  nlp-augmenter:
    image: registry.openrainbow.org/rainbow/nlp-augmentation-service:0.0.1-SNAPSHOT
    networks: ['rasa-network']
    restart: always
    ports:
      - 50051:50051


volumes:
  mongodbdata:
    driver: local


networks: {rasa-network: {}}

any idea please ?
